src/System.Net.Mime/src/System/Net/Mime/TrackingValidationObjectDictionary.cs

```diff
@@ -0,0 +1,198 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Specialized;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace System.Net
+{
+    // TrackingValidationObjectDictionary uses an internal collection of objects to store
+    // only those objects which are not strings.  It still places a copy of the string
+    // value of these objects into the base StringDictionary so that the public methods
+    // of StringDictionary still function correctly.  
+    // NOTE:  all keys are converted to lowercase prior to adding to ensure consistency of 
+    // values between  keys in the StringDictionary and internalObjects because StringDictionary
+    // automatically does this internally
+    internal sealed class TrackingValidationObjectDictionary : StringDictionary
+    {
+        #region Private Fields
+
+        // even though validators may exist, we should not initialize this initially since by default it is empty
+        // and it may never be populated with values if the user does not set them
+        private readonly IDictionary<string, ValidateAndParseValue> _validators;
+        private IDictionary<string, object> _internalObjects = null;
+
+        #endregion
+
+        #region Constructors
+
+        // it is valid for validators to be null.  this means that no validation should be performed
+        internal TrackingValidationObjectDictionary(IDictionary<string, ValidateAndParseValue> validators)
+        {
+            IsChanged = false;
+            _validators = validators;
+        }
+
+        #endregion
+
+        #region Private Methods
+
+        // precondition:  key must not be null
+        // addValue determines if we are doing a set (false) or an add (true)
+        private void PersistValue(string key, string value, bool addValue)
+        {
+            Debug.Assert(key != null, "key was null");
+
+            // StringDictionary will automatically store the key as lower case so
+            // we must convert it so that the validators and internalObjects will
+            // be consistent
+            key = key.ToLowerInvariant();
```

Comment:
`StringDictionary` uses `ToLowerInvariant()` on the key, so this means we're doing the ToLower allocation twice... It'd be nice to avoid. Any way to use `StringComparer.InvariantCultureIgnoreCase` as the comparer for `validators` and `internalObjects`?